### PR TITLE
QA file-restructure

### PR DESCRIPTION
**Pull request by mixmix**
- a script for testing CLI
- some gotchas!
- WIP bugs!

PLEASE review + merge https://github.com/entropyxyz/cli/pull/250 first (makes this PR smaller)

### CODE REVIEW

File: tests/cli.test.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#! /usr/bin/bash
```
Comment (mixmix) [inline tests/cli.test.sh:1]: Here's a script I've been using to QA the CLI

It's breaking down at Transfer :angry:

File: src/transfer/command.ts [hunk 2]
```diff
@@ -1,20 +1,31 @@
...
     .action(async (destination, amount, opts) => {
+      console.log({ destination, amount, opts })
+
+      // TODO: destination as <name|address> ?
       const entropy = await loadEntropy(opts.account, opts.endpoint)
+        .catch(err => {
+          // WIP here. SOMETHING is wrecking the config upstream
+          console.error("loadEntropy failed", err)
+          throw err
+        })
```
Comment (mixmix) [inline src/transfer/command.ts:23]: something :fire:

File: src/transfer/command.ts [hunk 3]
```diff
@@ -1,20 +1,31 @@
 import { Command } from "commander"
-import { accountOption, endpointOption, loadEntropy, passwordOption } from "src/common/utils-cli"
+import { accountOption, endpointOption, loadEntropy, passwordOption } from "../common/utils-cli"
 import { EntropyTransfer } from "./main"
 
 export function entropyTransferCommand () {
-  const transferCommand = new Command('tranfer')
+  const transferCommand = new Command('transfer')
```
Comment (mixmix) [inline src/transfer/command.ts:6]: _spelling!_

File: tests/cli.test.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+print "// BALANCE ///////////////////////////////////////////////// "
+
+print "balance naynay"
+entropy balance naynay
+
+print "balance 5CqJyjALDFz4sKjQgK8NXBQGHCWAiV63xXn2Dye393Y6Vghz"
+# entropy balance faucet
+entropy balance 5CqJyjALDFz4sKjQgK8NXBQGHCWAiV63xXn2Dye393Y6Vghz
+
+
```
Comment (mixmix) [inline tests/cli.test.sh:42]: maybe we should log the config through-out this script to see where it done breaks

File: src/account/command.ts [hunk 5]
```diff
@@ -73,8 +78,14 @@ function entropyAccountList () {
...
-      const storedConfig = await config.get()
-      const accounts = EntropyAccount.list(storedConfig)
+      const accounts = await config.get()
+        .then(storedConfig => EntropyAccount.list(storedConfig))
+        .catch((err) => {
+          if (err.message.includes('currently no accounts')) return []
+
+          throw err
+        })
+
```
Comment (mixmix) [inline src/account/command.ts:88]: `entropy account ls`
If I don't have any accounts yet I don't expect an error, I expect `[]`

File: src/account/command.ts [hunk 6]
```diff
@@ -35,7 +38,8 @@ function entropyAccountCreate () {
 
       cliWrite({
         name: newAccount.name,
-        address: newAccount.address
+        address: newAccount.address,
+        verifyingKeys: []
```
Comment (mixmix) [inline src/account/command.ts:42]: `entropy account create`
`entropy account create`

Have them output the same format as `entropy account ls`
:heavy_check_mark: consistency
:heavy_check_mark: set up expectations that there is something interesting/ important meant to go in here

File: src/account/main.ts [hunk 7]
```diff
@@ -19,28 +19,28 @@ export class EntropyAccount extends EntropyBase {
...
   static async import ({ name, seed, path }: AccountImportParams ): Promise<EntropyAccountConfig> {
-    // WARNING: #create currently depends on this => be careful modifying this function
-
     await wasmGlobalsReady()
     const keyring = new Keyring({ seed, path, debug: true })
+
     const fullAccount = keyring.getAccount()
     // TODO: sdk should create account on constructor
-    const { admin } = keyring.getAccount()
+    const data = fixData(fullAccount)
```
Comment (mixmix) [inline src/account/main.ts:29]: There is a bug in the SDK where you are "cloning" with `JSON.parse(JSON.stringify(obj)) and it fucks up the Uint8Array, which then fucks up config, logging etc.

This is a quick hack

File: src/account/main.ts [hunk 8]
```diff
@@ -49,7 +49,7 @@ export class EntropyAccount extends EntropyBase {
     return accounts.map((account: EntropyAccountConfig) => ({
       name: account.name,
       address: account.address,
-      verifyingKeys: account?.data?.admin?.verifyingKeys
+      verifyingKeys: account?.data?.registration?.verifyingKeys || []
```
Comment (mixmix) [inline src/account/main.ts:52]: :fire: CHECK?
I assume `verifyingKeys` were meant to go under `registration` and not `admin` right?

File: src/common/masking.ts [hunk 9]
```diff
@@ -1,37 +1,38 @@
...
-const DEFAULT_MASKED_FIELDS = [
+const PREFIX = 'data:application/UI8A;base64,'
+const DEFAULT_MASKED_FIELDS = new Set([
   'seed',
   'secretKey',
   'addressRaw',
-];
+]);
 
 export function maskPayload (payload: any): any {
```
Comment (mixmix) [inline src/common/masking.ts:8]: I ended up in here cos of the masking of the badly handled UInt8Arrays... and reading it I saw there was not test and it was pretty confusing, so I wrote tests and wrote a simpler function with less dependencies.

I think it's an improvement ... I mean it needed to check the `DEFAULT_MASKED_FIELDS` earlier cos the test for "is this a Uint8Array or a munted Uint8Array (object like `{ "0": 34, "1": 23, ... }`)  ... yeah that broke when I fixed the bad keys coming from SDK...

There's a chance I introduced more edge cases...but look at the tests I guess!

File: src/common/utils-cli.ts [hunk 10]
```diff
@@ -20,7 +20,7 @@ function getConfigOrNull () {
 
 export function endpointOption () {
   return new Option(
-    '-e, --endpoint <endpoint>',
+    '-e, --endpoint <url>',
```
Comment (mixmix) [inline src/common/utils-cli.ts:23]: reading the docs the pattern is  `-shortcode --longcode <input-type>`

File: src/common/utils-cli.ts [hunk 11]
```diff
@@ -39,41 +39,48 @@ export function endpointOption () {
...
     .env('ENTROPY_ACCOUNT')
-    .argParser(async (account) => {
-      if (storedConfig && storedConfig.selectedAccount !== account) {
-        // Updated selected account in config with new address from this option
-        await config.set({
-          ...storedConfig,
-          selectedAccount: account
-        })
-      }
+    .argParser(addressOrName => {
```
Comment (mixmix) [inline src/common/utils-cli.ts:64]: :fire: :fire: NASTY bug

this thing don't handle `async`!

See comments below

Comment (mixmix) [reply src/common/utils-cli.ts:64]: I caught this cos I tried to register `naynay` in my QA and it was registering `faucet`!

File: src/config/encoding.ts [hunk 12]
```diff
@@ -1,12 +1,18 @@
...
 }
 
-export function deserialize (config) {
-  return JSON.parse(config, reviver)
+export function deserialize (config: string) {
+  try {
+    return JSON.parse(config, reviver)
+  } catch (err) {
+    console.log('broken config:', config)
+    // WIP here: nothing being passed in?!!
```
Comment (mixmix) [inline src/config/encoding.ts:13]:
```suggestion
```

File: src/config/index.ts [hunk 13]
```diff
@@ -73,6 +73,20 @@ export async function set (config: EntropyConfig, configPath = CONFIG_PATH) {
...
+  // no need for update
+
+  const newConfig = {
+    ...storedConfig,
+    selectedAccount: account.name
+  }
+  await set(newConfig, configPath)
+  return newConfig
+}
+
```
Comment (mixmix) [inline src/config/index.ts:89]: :fire: one bug I saw was inconsistency around what `config.selectedAccount` was (an address? a name?).

I decided to fix it to be `name` because then in the `--help` you will see the default by `name` not some crazy address

I then changed
- `accountOption` to deal with `name`
- added a special method `setSelectedAccount` to handle the persistence of the right thing

Things I'm not happy about:
- we are now doing a LOT of get/set/get/set of the same config in a few places
- some places we need to persist the whole config (e.g. new `verifyingKeys`... and so the knowledge of what the `selecttedAccount` is is leaking out into the code)
- does this suggest there should be a `getSelectedAccount` ?? where do the helper methods stop!

File: tests/qa.sh [hunk 14]
```diff
@@ -0,0 +1,88 @@
+#! /usr/bin/bash
+
+# WARNING - this script nukes your config!
+#
+# Dependencies
+#   - internet connection
+#   - jq - see https://jqlang.github.io/jq
+#
+# Run
+#   $ yarn build && ./tests/qa.sh
```
Comment (mixmix) [inline tests/qa.sh:10]: I'm pretty stoked with this. This script allowed me to hammer the CLI repeatedly till I got all the weird lil bugs out.

I imagine this could be the basis for a CLI integration test that uses frankie's new CLI runner in the future so we can check the outputs are good

File: tests/qa.sh [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+entropy sign -a naynay "some content!\nNICE&SIMPLE"
+
+
+
+print "// PROGRAM /////////////////////////////////////////////////"
+
+DATE=`date` && echo "wasm junk - ${DATE}" > /tmp/entropy.fake.wasm
+echo '{"type": "object"}' > /tmp/entropy.configSchema.fake.json
+echo '{"type": "object"}' > /tmp/entropy.auxDataSchema.fake.json
+print "program deploy"
```
Comment (mixmix) [inline tests/qa.sh:84]:
```suggestion
print "program deploy"
DATE=`date` && echo "wasm junk - ${DATE}" > /tmp/entropy.fake.wasm
echo '{"type": "object"}' > /tmp/entropy.configSchema.fake.json
echo '{"type": "object"}' > /tmp/entropy.auxDataSchema.fake.json
```

**Comment by mixmix** [comment]
# Checklist

- [ ] fix failing CI
- [ ] deal with argParser / default collision
    - reproduce by running `./tests/qa.sh` , you should see naynay get the verifyingKeys on register... but it's faucet
    - this happens cos `Option.argParser` + `Option.default` colliding / fighting? 
    - See  https://github.com/tj/commander.js/issues/1641

File: src/account/main.ts (hunk 8 above)
Comment (mixmix) [reply src/account/main.ts:52]: AH tests say no. I don't know what the difference is between `admin + registration.
I changed where vks are being pushed into elsewhere in code too

File: src/common/masking.ts (hunk 9 above)
Comment (mixmix) [reply src/common/masking.ts]: Extracted into another PR

File: src/common/utils-cli.ts (hunk 11 above)
Comment (mixmix) [reply src/common/utils-cli.ts:65]: Resolved in another PR

File: src/config/index.ts (hunk 13 above)
Comment (mixmix) [reply src/config/index.ts:89]: Moved to another PR

File: src/account/main.ts (hunk 8 above)
Comment (mixmix) [inline src/account/main.ts:52]: This was a bug I believe

Comment (mixmix) [reply src/account/main.ts:52]: changed tests

**Comment by rh0delta** [comment]
@mixmix similar for this PR, need to rebase off of current version of dev.

**Comment by mixmix** [comment]
@rh0delta DONE!

File: src/config/types.ts [hunk 16]
```diff
@@ -43,6 +43,7 @@ export interface EntropyAccount {
 
 export enum EntropyAccountContextType {
   programDev = 'PROGRAM_DEV_KEY',
+  // QUESTION: what is this and why is registration = ADMIN_KEY?!
```
[the remark this answers is not in the document]
Comment (mixmix) [reply src/config/types.ts:46]: In future, user should be able to set their own program mod key, at the moment it's not possible

[the remark this answers is not in the document]
Comment (mixmix) [reply src/config/types.ts:46]:
```suggestion
```

**Review by rh0delta — APPROVED** [review]
